apt-hook/hook.cc

```diff
@@ -201,69 +282,230 @@ int main(int argc, char *argv[])
+                     ngettext("%d of the updates is from UA Infra: ESM.",
+                              "%d of the updates are from UA Infra: ESM.",
+                              res.enabled_esms_i),
+                     res.enabled_esms_i);
+            ioprintf(std::cout, "\n");
+         }
+         if (res.enabled_esms_a > 0)
+         {
+            ioprintf(std::cout,
+                     ngettext("%d of the updates is from UA Apps: ESM.",
+                              "%d of the updates are from UA Apps: ESM.",
+                              res.enabled_esms_a),
+                     res.enabled_esms_a);
+            ioprintf(std::cout, "\n");
+         }
       }
-      if (res.disabled_esms_a > 0)
+
+      if (res.disabled_esms_i > 0 || res.disabled_esms_a > 0)
       {
-         ioprintf(std::cout,
-                  ngettext("%d additional update is available with UA Apps: ESM.",
-                           "%d additional updates are available with UA Apps: ESM.",
-                           res.disabled_esms_a),
-                  res.disabled_esms_a);
+         if (command_used != "update")
+            std::cout << std::endl;
+         if (res.disabled_esms_i > 0)
+         {
+            ioprintf(std::cout,
+                     ngettext("%d additional update is available with UA Infra: ESM.",
+                              "%d additional updates are available with UA Infra: ESM.",
+                              res.disabled_esms_i),
+                     res.disabled_esms_i);
+            ioprintf(std::cout, "\n");
+         }
+         if (res.disabled_esms_a > 0)
+         {
+            ioprintf(std::cout,
+                     ngettext("%d additional update is available with UA Apps: ESM.",
+                              "%d additional updates are available with UA Apps: ESM.",
+                              res.disabled_esms_a),
+                     res.disabled_esms_a);
+            ioprintf(std::cout, "\n");
+         }
+
+         ioprintf(std::cout, gettext("To see these additional updates run: apt list --upgradable"));
+         ioprintf(std::cout, "\n");
+         ioprintf(std::cout, gettext("See https://ubuntu.com/advantage or run: sudo ua status"));
          ioprintf(std::cout, "\n");
       }
-
-      ioprintf(std::cout, gettext("To see these additional updates run: apt list --upgradable"));
-      ioprintf(std::cout, "\n");
-      ioprintf(std::cout, gettext("See https://ubuntu.com/advantage or run: sudo ua status"));
-      ioprintf(std::cout, "\n");
+   } else if (subcommand == PreInvoke) {
+      if (command_used == "upgrade" || command_used == "dist-upgrade") {
+         // Print static message if present. Used for "Expiring soon" and "Expired but in grace period" messages.
+         std::ifstream message_file(CONTRACT_EXPIRY_STATUS_MESSAGE_PATH);
+         if (message_file.is_open()) {
+            std::cout << message_file.rdbuf();
+            message_file.close();
+         }
```

Comment:
@orndorffgrant I was expecting this to behave the same way as other templates: lib/ua_update_messages will emit `/var/lib/ubuntu-advantage/messages/contract-expiry-status.tmpl` and expect apt hook will update and write `/var/lib/ubuntu-advantage/messages/contract-expiry-status`

```suggestion
         process_template_file(
            CONTRACT_EXPIRY_STATUS_MESSAGE_TEMPLATE_PATH,
            CONTRACT_EXPIRY_STATUS_MESSAGE_TEMPLATE,
            esm_a_packages_count,
            space_separated_esm_a_packages,
            esm_i_packages_count,
            space_separated_esm_i_packages
         );
```

Reply:
Yep that makes sense! Then this won't be a special case.